File: cloudflare_worker/src/catalog.py
"""Catalog-record sanitization helpers (js-free sibling module).

These pure input-validation helpers -- string trimming, path-segment
validation, and the public catalog-record builder -- were split out of the
11k-line entry.py so the sanitization layer is one small, scannable module.
The Worker runtime bundles sibling modules in src/, so entry.py's re-import
resolves both on Cloudflare and in the test suite (which parses this file
the same way it parses entry.py).
"""

import re
from urllib.parse import unquote

MAX_REPO_SEGMENT = 80
ROOM_NAME_RE = re.compile(r"^[A-Za-z0-9._:-]+$")


def safe_segment(value, max_length=MAX_REPO_SEGMENT):
    value = unquote(value).strip()
    if not value or len(value) > max_length:
        return None
    if not ROOM_NAME_RE.match(value):
        return None
    return value


def clean_string(value, max_length=240):
    if not isinstance(value, str):
        return ""
    return value.strip()[:max_length]


def safe_contribution_transport(data):
    """Return bounded optional snapshot fields without adding them to a record."""
    if not isinstance(data, dict):
        return {"present": False, "payload": "", "signature": "", "warning": ""}
    present = "contributionPayload" in data or "contributionSig" in data
    if not present:
        return {"present": False, "payload": "", "signature": "", "warning": ""}
    payload = data.get("contributionPayload")
    signature = data.get("contributionSig")
    if not isinstance(payload, str) or not isinstance(signature, str):
        return {
            "present": True,
            "payload": "",
            "signature": "",
            "warning": "invalid_contribution_transport",
        }
    if len(payload) > 64 * 1024:
        return {
            "present": True,
            "payload": "",
            "signature": "",
            "warning": "contribution_payload_too_large",
        }
    if not payload or not signature or len(signature) > 200:
        return {
            "present": True,
            "payload": "",
            "signature": "",
            "warning": "invalid_contribution_transport",
        }
    return {
        "present": True,
        "payload": payload,
        "signature": signature,
        "warning": "",
    }


def clean_int_series(value, length=52, max_value=1000000):
    if not isinstance(value, list):
        return [0] * length
    series = []
    for item in value[-length:]:
        try:
            number = int(item)
        except (TypeError, ValueError):
            number = 0
        series.append(max(0, min(number, max_value)))
    return ([0] * max(0, length - len(series))) + series


def clean_optional_integer(value, max_value):
    """Return a bounded integer metric, or None when it was not shared."""
    if isinstance(value, bool) or not isinstance(value, (int, float)):
        return None
    if isinstance(value, float) and (
            value != value or value in (float("inf"), float("-inf"))
            or not value.is_integer()):
        return None
    if value < 0:
        return None
    return min(int(value), max_value)


def clean_optional_usage(used_value, total_value):
    """Normalize a used/total byte pair without manufacturing partial data."""
    maximum = 1 << 50
    used = clean_optional_integer(used_value, maximum)
    total = clean_optional_integer(total_value, maximum)
    if used is None or total is None or total <= 0:
        return None, None
    return min(used, total), total


def clean_logo_metadata(value):
    """Keep only bounded, content-free inputs for native logo generation."""
    value = value if isinstance(value, dict) else {}

    def labels(name, limit, max_length):
        raw = value.get(name)
        if not isinstance(raw, list):
            return []
        output = []
        for item in raw:
            label = clean_string(item, max_length)
            if label:
                output.append(label)
            if len(output) >= limit:
                break
        return output

    languages = {}
    raw_languages = value.get("languages")
    if isinstance(raw_languages, dict):
        for raw_name, raw_bytes in raw_languages.items():
            name = clean_string(raw_name, 80)
            if not name:
                continue
            try:
                byte_count = int(raw_bytes)
            except (TypeError, ValueError):
                byte_count = 0
            languages[name] = max(0, min(byte_count, 1 << 50))
            if len(languages) >= 12:
                break

    return {
        "description": clean_string(value.get("description", ""), 500),
        "languages": languages,
        "topics": labels("topics", 12, 80),
        "fileStructure": labels("fileStructure", 24, 120),
        "frameworks": labels("frameworks", 12, 80),
        "projectCategory": clean_string(
            value.get("projectCategory", ""), 80),
    }


def safe_catalog_record(data):
    if not isinstance(data, dict):
        return None

    owner = safe_segment(data.get("owner", ""))
    name = safe_segment(data.get("name", ""))
    public_key = clean_string(data.get("maintainer", ""), 120)
    if not owner or not name or not public_key:
        return None

    now = clean_string(data.get("updatedAt", ""), 32)
    # Visibility fails closed. Older/malformed publishers that omit the field
    # may hide a public repository until they republish, but they can never make
    # a private repository discoverable by accident.
    visibility = (
        "public" if data.get("visibility") == "public" else "private"
    )
    # On-disk mirror size (bytes) the publishing node reports. Clamped to a sane
    # non-negative integer; 0 when absent or unparseable. Drives the size figures
    # and "data hosted" leaderboards on the network page.
    try:
        size_bytes = max(0, min(int(data.get("sizeBytes", 0) or 0), 1 << 50))
    except (TypeError, ValueError):
        size_bytes = 0
    try:
        key_epoch = max(0, min(int(data.get("keyEpoch", 0) or 0), 1 << 31))
    except (TypeError, ValueError):
        key_epoch = 0
    solana = clean_string(data.get("solana", ""), 64)
    if not re.fullmatch(r"[1-9A-HJ-NP-Za-km-z]{32,44}", solana):
        solana = ""
    mem_used, mem_total = clean_optional_usage(
        data.get("memUsedBytes"), data.get("memTotalBytes"))
    disk_used, disk_total = clean_optional_usage(
        data.get("diskUsedBytes"), data.get("diskTotalBytes"))
    actions_fields = {"actionsEnabled", "actionsState"}.intersection(data)
    if actions_fields and actions_fields != {"actionsEnabled", "actionsState"}:
        return None
    actions_enabled = data.get("actionsEnabled")
    actions_state = data.get("actionsState")
    if actions_fields and (
        not isinstance(actions_enabled, bool)
        or (
            (not actions_enabled and actions_state != "disabled")
            or (actions_enabled and actions_state not in {"enabled", "running"})
        )
    ):
        return None
    record = {
        "owner": owner,
        "name": name,
        "visibility": visibility,
        "mirrorEncryption": (
            "owner-sealed-v1"
            if data.get("mirrorEncryption") == "owner-sealed-v1" else ""),
        "opaqueRepoId": clean_string(data.get("opaqueRepoId", ""), 64).lower(),
        "keyEpoch": key_epoch,
        "encryptedManifestHash": clean_string(
            data.get("encryptedManifestHash", ""), 64).lower(),
        "encryptedManifestSig": clean_string(
            data.get("encryptedManifestSig", ""), 220),
        "sizeBytes": size_bytes,
        "description": clean_string(data.get("description", ""), 240),
        "logoMetadata": clean_logo_metadata(data.get("logoMetadata")),
        "cloneUrl": clean_string(data.get("cloneUrl", ""), 2048),
        "solana": solana,
        "channel": clean_string(data.get("channel", f"#{owner}-{name}"), 120),
        "hostedSince": clean_string(data.get("hostedSince", ""), 32),
        "lastSync": clean_string(data.get("lastSync", ""), 32),
        "updatedAt": now,
        # Stable identity (first/root commit) shared by every mirror of this repo,
        # so the network page can group mirrors under different owners into one
        # card. Falls back to the repo name on the website when absent.
        "rootCommit": clean_string(data.get("rootCommit", ""), 64),
        "source": clean_string(data.get("source", "local-node"), 40),
        # Node facts mirrored from the publishing node's live advert, so the Mirror
        # nodes view can show latest commit / issues / platform / version / id for a
        # node even while it's offline (adhoc #56). Point-in-time, like sizeBytes.
        "commit": clean_string(data.get("commit", ""), 64),
        "branch": clean_string(data.get("branch", ""), 120),
        "issueCount": clean_string(data.get("issueCount", ""), 12),
        # Highest issue number ever assigned (closed/deleted included), so the
        # relay can propose the desktop's real next number for a ForkBot issue
        # (see _forkbot_next_issue_number). Distinct from issueCount, which is
        # only the open count.
        "issueMaxNumber": clean_string(data.get("issueMaxNumber", ""), 12),
        "commitCount": clean_string(data.get("commitCount", ""), 12),
        "branchCount": clean_string(data.get("branchCount", ""), 12),
        "pullCount": clean_string(data.get("pullCount", ""), 12),
        "discussionCount": clean_string(data.get("discussionCount", ""), 12),
        "activityWeeks": clean_int_series(data.get("activityWeeks"), 52),
        "worktreeCount": clean_string(data.get("worktreeCount", ""), 12),
        "artifactCount": clean_string(data.get("artifactCount", ""), 12),
        "platform": clean_string(data.get("platform", ""), 16),
        "version": clean_string(data.get("version", ""), 32),
        "nodeId": clean_string(data.get("nodeId", ""), 64),
        # How many clones and website (browse/fetch) requests this node has served
        # for the repo. Purely local counters otherwise, mirrored here so the Mirror
        # nodes view can show a node's contribution even while it's offline.
        "clonesServed": clean_string(data.get("clonesServed", ""), 12),
        "websiteServed": clean_string(data.get("websiteServed", ""), 12),
        "maintainer": public_key,
        "signature": clean_string(data.get("signature", ""), 220),
        # Owner-signed fingerprint of the repo's served refs (sha256 over the
        # canonical heads+tags advertisement; see advertised_refs_canonical). The
        # relay pins this and refuses to serve any mirror whose live advertisement
        # doesn't hash to it — so a tampered or rolled-back mirror can't be cloned.
        "stateHash": clean_string(data.get("stateHash", ""), 64),
        "stateSig": clean_string(data.get("stateSig", ""), 220),
    }
    # Keep this pair absent on legacy records so their catalog-v2 signatures
    # still verify. New reports are a strict, signed capability/status only:
    # arbitrary Actions configuration never reaches the stored public record.
    if actions_fields:
        record["actionsEnabled"] = actions_enabled
        record["actionsState"] = actions_state
    # Keep absent telemetry absent (rather than adding null fields) so a
    # catalog-v2 signature produced by an older, opted-out client continues to
    # verify after this schema extension. Consumers still expose unknown values
    # as null in their response shape.
    cpu_percent = clean_optional_integer(data.get("cpuPercent"), 100)
    if cpu_percent is not None:
        record["cpuPercent"] = cpu_percent
    if mem_total is not None:
        record["memUsedBytes"] = mem_used
        record["memTotalBytes"] = mem_total
    if disk_total is not None:
        record["diskUsedBytes"] = disk_used
        record["diskTotalBytes"] = disk_total
    return record

File: mirror_node/catalog.go
package mirrornode

import (
	"bytes"
	"context"
	"crypto/sha256"
	"encoding/hex"
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"os"
	"os/exec"
	"path/filepath"
	"runtime"
	"strconv"
	"strings"
	"syscall"
	"time"
)

type CatalogPublisher struct {
	URL      string
	Owner    string
	Node     string
	Version  string
	Identity *Identity
	Client   *http.Client
}

func (p *CatalogPublisher) Publish(ctx context.Context, repo Repository, stateHash string) error {
	record, err := p.record(ctx, repo, stateHash)
	if err != nil {
		return err
	}
	body, err := json.Marshal(record)
	if err != nil {
		return err
	}
	req, err := http.NewRequestWithContext(ctx, http.MethodPost, p.URL, bytes.NewReader(body))
	if err != nil {
		return err
	}
	req.Header.Set("Content-Type", "application/json")
	req.Header.Set("Accept", "application/json")
	req.Header.Set("User-Agent", "ForkMesh-Mirror-Node/"+p.Version)
	client := p.Client
	if client == nil {
		client = &http.Client{Timeout: 30 * time.Second}
	}
	response, err := client.Do(req)
	if err != nil {
		return err
	}
	defer response.Body.Close()
	if response.StatusCode < 200 || response.StatusCode >= 300 {
		body, _ := io.ReadAll(io.LimitReader(response.Body, 1024))
		return fmt.Errorf("catalog returned HTTP %d: %s", response.StatusCode,
			boundedText(body, 240))
	}
	return nil
}

func (p *CatalogPublisher) record(ctx context.Context, repo Repository, stateHash string) (map[string]any, error) {
	branch := gitValue(ctx, repo.GitDir, "symbolic-ref", "--short", "HEAD")
	commit := gitValue(ctx, repo.GitDir, "rev-parse", "HEAD")
	if branch == "" || commit == "" || !digestPattern.MatchString(stateHash) {
		return nil, fmt.Errorf("repository head or state hash unavailable")
	}
	now := strconv.FormatInt(time.Now().UnixMilli(), 10)
	rootCommit := firstLine(gitValue(ctx, repo.GitDir, "rev-list", "--max-parents=0", "HEAD"))
	commitCount := gitValue(ctx, repo.GitDir, "rev-list", "--count", "HEAD")
	branchCount := countLines(gitValue(ctx, repo.GitDir, "for-each-ref", "--format=%(refname)", "refs/heads"))
	paths := gitLines(ctx, repo.GitDir, "ls-tree", "-r", "--name-only", "HEAD", "--", ".forkmesh")
	issueCount, issueMax, pullCount, discussionCount, artifacts := catalogPathCounts(paths)
	size := directorySize(repo.GitDir)
	memUsed, memTotal := memoryUsage()
	diskUsed, diskTotal := diskUsage(repo.GitDir)
	record := map[string]any{
		"owner": p.Owner, "name": repo.Name, "visibility": "public",
		"mirrorEncryption": "", "opaqueRepoId": "", "keyEpoch": 0,
		"encryptedManifestHash": "", "encryptedManifestSig": "",
		"sizeBytes": size, "description": "",
		"logoMetadata": map[string]any{"description": "", "languages": map[string]int64{}, "topics": []string{}, "fileStructure": []string{}, "frameworks": []string{}, "projectCategory": ""},
		"cloneUrl":     "", "solana": "", "channel": "#" + p.Owner + "-" + repo.Name,
		"hostedSince": "", "lastSync": now, "updatedAt": now,
		"rootCommit": rootCommit, "source": "remote-clone", "commit": commit, "branch": branch,
		"issueCount": strconv.Itoa(issueCount), "issueMaxNumber": strconv.Itoa(issueMax),
		"commitCount": commitCount, "branchCount": strconv.Itoa(branchCount),
		"pullCount": strconv.Itoa(pullCount), "discussionCount": strconv.Itoa(discussionCount),
		"activityWeeks": make([]int, 52), "worktreeCount": "0", "artifactCount": strconv.Itoa(artifacts),
		"platform": runtime.GOOS, "version": p.Version, "nodeId": "",
		"clonesServed": "0", "websiteServed": "0", "maintainer": p.Identity.PublicKey(),
		"signature": "", "stateHash": stateHash, "stateSig": "",
		"cpuPercent": cpuPercent(), "memUsedBytes": memUsed, "memTotalBytes": memTotal,
		"diskUsedBytes": diskUsed, "diskTotalBytes": diskTotal,
		"machineName": p.Node, "runtimeMode": "headless",
	}
	statePayload := "forkmesh-repostate-v1\n" + p.Owner + "\n" + repo.Name + "\n" + stateHash + "\n" + now
	record["stateSig"] = p.Identity.Sign([]byte(statePayload))
	signed := make(map[string]any, len(record)-1)
	for key, value := range record {
		if key != "signature" {
			signed[key] = value
		}
	}
	canonical, err := json.Marshal(signed)
	if err != nil {
		return nil, err
	}
	digest := sha256.Sum256(canonical)
	record["catalogSigVersion"] = 2
	record["catalogSig"] = p.Identity.Sign([]byte("forkmesh-catalog-v2\n" + hex.EncodeToString(digest[:])))
	return record, nil
}

func gitValue(ctx context.Context, gitDir string, args ...string) string {
	base := append([]string{"--git-dir=" + gitDir}, args...)
	out, err := exec.CommandContext(ctx, "git", base...).Output()
	if err != nil {
		return ""
	}
	return strings.TrimSpace(string(out))
}

func gitLines(ctx context.Context, gitDir string, args ...string) []string {
	value := gitValue(ctx, gitDir, args...)
	if value == "" {
		return nil
	}
	return strings.Split(value, "\n")
}

func countLines(value string) int {
	if strings.TrimSpace(value) == "" {
		return 0
	}
	return len(strings.Split(strings.TrimSpace(value), "\n"))
}

func firstLine(value string) string {
	if index := strings.IndexByte(value, '\n'); index >= 0 {
		value = value[:index]
	}
	if len(value) > 64 {
		value = value[:64]
	}
	return strings.TrimSpace(value)
}

func catalogPathCounts(paths []string) (issues, maxIssue, pulls, discussions, artifacts int) {
	for _, path := range paths {
		path = strings.TrimSpace(path)
		if strings.HasPrefix(path, ".forkmesh/issues/open/") && strings.HasSuffix(path, ".json") {
			issues++
		}
		if strings.HasPrefix(path, ".forkmesh/issues/") {
			for _, part := range strings.Split(path, "/") {
				if value, err := strconv.Atoi(strings.TrimPrefix(strings.TrimSuffix(part, ".json"), "issue-")); err == nil && value > maxIssue {
					maxIssue = value
				}
			}
		}
		if strings.HasPrefix(path, ".forkmesh/pulls/open/") && strings.HasSuffix(path, ".json") {
			pulls++
		}
		if strings.HasPrefix(path, ".forkmesh/discussions/") && strings.HasSuffix(path, ".json") {
			discussions++
		}
		if strings.HasPrefix(path, ".forkmesh/artifacts/") {
			artifacts++
		}
	}
	return
}

func directorySize(root string) int64 {
	var total int64
	_ = filepathWalk(root, func(info os.FileInfo) {
		if info.Mode().IsRegular() {
			total += info.Size()
		}
	})
	return total
}

var filepathWalk = func(root string, visit func(os.FileInfo)) error {
	return filepath.Walk(root, func(_ string, info os.FileInfo, err error) error {
		if err == nil {
			visit(info)
		}
		return nil
	})
}

func memoryUsage() (uint64, uint64) {
	var info syscall.Sysinfo_t
	if syscall.Sysinfo(&info) != nil {
		return 0, 0
	}
	total := info.Totalram * uint64(info.Unit)
	free := (info.Freeram + info.Bufferram) * uint64(info.Unit)
	return total - free, total
}

func diskUsage(path string) (uint64, uint64) {
	var info syscall.Statfs_t
	if syscall.Statfs(path, &info) != nil {
		return 0, 0
	}
	total := info.Blocks * uint64(info.Bsize)
	free := info.Bavail * uint64(info.Bsize)
	return total - free, total
}

func cpuPercent() int { return 0 }
